tests/quality/test_quality.py

```diff
@@ -0,0 +1,143 @@
+        numerical_transformer = NumericalTransformer(null_column=False)
+        target = numerical_transformer.fit_transform(target, list(target.columns))
+        target = format_array(target)
+        scores = []
+        for transformer in transformers:
+            ht = HyperTransformer(data_type_transformers={data_type: transformer})
+            ht.fit(features)
+            transformed_features = ht.transform(features).to_numpy()
+            scores.append(get_regression_score(transformed_features, target))
+
+        all_scores[column] = pd.Series(scores, index=index)
+
+    return all_scores
+
+
+def validate_relative_score(scores, transformer):
+    scores_without_transformer = scores.drop(transformer)
+    if len(scores_without_transformer) == 0:
+        return
+
+    means = scores_without_transformer.mean()
+    means_above_threshold = means > THRESHOLD
+    standard_deviations = scores.std()
+    minimum_scores = means[means_above_threshold] - 2 * standard_deviations[means_above_threshold]
+
+    assert all(scores.loc[transformer, means_above_threshold] >= minimum_scores)
+
+
+def get_test_cases():
+    test_cases = []
+    path = os.path.join(os.path.dirname(__file__), 'dataset_info.csv')
+    datasets = pd.read_csv(path)
+    for _, row in datasets.iterrows():
+        if row['table_size'] < MAX_SIZE and row['modality'] in ['single-table', 'timeseries']:
+            for data_type in eval(row['table_types']):
+                if data_type not in TYPES_TO_SKIP:
+                    test_cases.append((data_type, row['name'], row['table_name']))
+
+    return test_cases
+
+
+def test_quality(subtests):
+    """Run all the quality test cases.
+
+    This test goes through each test case and tests all the transformers
+    of the test case's type against the test case's dataset. First, all
+    the transformers of the data type are used to transform every column
+    in the dataset of that type. A regression model is then trained on
+    those transformed column to try and predict every numerical column
+    in the dataset. The scores for each transformer are then compared to
+    the mean score of the rest of the transformers to make sure none are more
+    than one standard deviation away.
+    """
+    transformers_by_type = get_transformers_by_type()
+    test_cases = get_test_cases()
+    data_types_to_test = [
+        data_type
+        for data_type in transformers_by_type.keys()
+        if data_type not in TYPES_TO_SKIP
+    ]
+    tested_data_types = dict.fromkeys(data_types_to_test, False)
+
+    for data_type, dataset_name, table_name in test_cases:
+        transformers = transformers_by_type[data_type]
+        (data, metadata) = download_single_table(dataset_name, table_name)
+        scores = get_transformer_scores(data, data_type, transformers, metadata)
+        if any(scores.mean() > THRESHOLD):
+            tested_data_types[data_type] = True
+
+        for transformer in scores.index:
```

Comment:
This loop should be done after the one above is finished.
I think the logic should be made a bit more similar to the performance tests one, where we:
1. Identify the transformers and datasets that need to be run
2. Evaluate and accumulate all the scores (loop above)
3. Validate the scores for the target transformer (this loop, possibly inside another function, running an ALL the accumulated results from point 2.)

Point 3 is the `subtest` level, where each `transformer` is a `test case`, and there are three possible outcomes for each one of them:
a. There are no valid datasets for the transformer data type, so the test fails.
b. The transformer is too bad when compared to the other datasets, so the test fails.
c. The transformer is good enough when compared to the other datasets, so the test passes.

Reply:
I actually think this approach may make things very messy unless we change the way the scores are computed/stored. Currently, each transformer has a set of datasets that scores are computed for. For each dataset, multiple scores are computed: one for every numeric column. We agreed on this because one transformer might be able to find correlations with one column, while a different transformer for the same data type finds correlations with another. In this case, both would be valid since they are each capable of producing data that exposes relationships.

The problem is that this makes step 2 hard, because all the scores cannot really be accumulated in one object. It isn't a table that just has the transformer type, dataset and score. There is a column for each target column's score, and those column names differ between datasets. Accumulating everything into one DataFrame would be a mess.

Maybe it could be a nested dict mapping {data_type: {dataset: scores}}. Then, for each transformer, we could get it's datasets and for each of those:
1. Validate at least one of the datasets has a high enough average score for at least one column
2. Compare the score for the transformer to the score of the others on those datasets